File: impl/typescript/src/account/builder.ts
/**
 * Fluent builders for NeuronAccount construction.
 *
 * Spec reference: 001 spec.md, contracts/account-builder.md
 *   - FR-001, FR-006: ParentAccountBuilder with publicKey + DID + currency.
 *   - FR-002, FR-007: ChildAccountBuilder with publicKey + parentPubKey + currency.
 *   - FR-007a: SharedAccountBuilder with multisigKey + currency.
 *   - FR-011: build() validates type-specific rules, throws on failure.
 *   - FR-011a: buildComplete() adds completeness checks (ledger attachment, registry binding).
 *   - FR-018: Optional ledger attachment on all builders.
 *   - FR-020: Currency symbol is mandatory on all accounts.
 *   - FR-022: RegistryBinding is mandatory for Child completeness.
 *   - FR-026: FeePayer is optional for Child accounts.
 *
 * Each builder follows the fluent pattern: `.withX().withY().build()`.
 * build() returns a readonly, immutable NeuronAccount instance.
 * buildComplete() enforces stricter completeness requirements.
 */

import type { NeuronPublicKey } from '../keylib/public-key.js';
import type { MultisigKey } from '../keylib/multisig-key.js';
import type { ParentAccount, ChildAccount, SharedAccount } from './account.js';
import type {
  NeuronDID,
  LedgerAttachment,
  RegistryBinding,
  LedgerAccountId,
} from './types.js';
import { AccountType } from './types.js';
import {
  missingRequiredField,
  invalidCurrencySymbol,
  accountIncomplete,
} from './errors.js';

/**
 * Fluent builder for Parent accounts.
 *
 * FR-001, FR-006: Requires publicKey, DID, and currencySymbol.
 * FR-018: Optional ledger attachment.
 *
 * Usage:
 *   const parent = new ParentAccountBuilder()
 *     .withPublicKey(pubKey)
 *     .withDID(did)
 *     .withCurrency('HBAR')
 *     .build();
 */
export class ParentAccountBuilder {
  private _publicKey?: NeuronPublicKey | undefined;
  private _did?: NeuronDID | undefined;
  private _currencySymbol?: string | undefined;
  private _ledgerAttachment?: LedgerAttachment | undefined;

  /**
   * Set the Parent's secp256k1 public key.
   * FR-006: MUST be a single NeuronPublicKey.
   */
  withPublicKey(pk: NeuronPublicKey): this {
    this._publicKey = pk;
    return this;
  }

  /**
   * Set the Parent's DID:key identifier.
   * FR-006, FR-012: MUST be present for Parent accounts.
   */
  withDID(did: NeuronDID): this {
    this._did = did;
    return this;
  }

  /**
   * Set the currency symbol.
   * FR-020: MUST be present on all account types.
   */
  withCurrency(symbol: string): this {
    this._currencySymbol = symbol;
    return this;
  }

  /**
   * Set the optional ledger attachment.
   * FR-018: MAY be present on any account type.
   */
  withLedgerAttachment(attachment: LedgerAttachment): this {
    this._ledgerAttachment = attachment;
    return this;
  }

  /**
   * Build a Parent account with basic validation.
   *
   * FR-011: Validates type-specific rules (V-PARENT-01..02).
   * Throws on the first missing required field.
   *
   * @returns Immutable ParentAccount instance
   * @throws AccountError NEURON-ACCT-002 if publicKey is missing
   * @throws AccountError NEURON-ACCT-002 if DID is missing
   * @throws AccountError NEURON-ACCT-008 if currencySymbol is missing or empty
   */
  build(): ParentAccount {
    if (this._publicKey == null) {
      throw missingRequiredField('publicKey is required for Parent account');
    }
    if (this._did == null) {
      throw missingRequiredField('did is required for Parent account');
    }
    if (this._currencySymbol == null || this._currencySymbol === '') {
      throw invalidCurrencySymbol('Currency symbol is required');
    }

    const account: ParentAccount = {
      accountType: AccountType.Parent,
      publicKey: this._publicKey,
      did: this._did,
      currencySymbol: this._currencySymbol,
      ...(this._ledgerAttachment != null
        ? { ledgerAttachment: this._ledgerAttachment }
        : {}),
    };

    return account;
  }

  /**
   * Build a complete Parent account with strict completeness validation.
   *
   * FR-011a: A Parent account is complete when it has all MUST fields
   * plus a ledger attachment.
   *
   * @returns Immutable ParentAccount instance with ledger attachment
   * @throws AccountError NEURON-ACCT-002 if required fields are missing
   * @throws AccountError NEURON-ACCT-008 if currencySymbol is missing
   * @throws AccountError NEURON-ACCT-007 if ledger attachment is missing
   */
  buildComplete(): ParentAccount {
    const account = this.build();
    if (account.ledgerAttachment == null) {
      throw accountIncomplete('Parent account requires ledger attachment for completeness');
    }
    return account;
  }
}

/**
 * Fluent builder for Child accounts.
 *
 * FR-002, FR-007: Requires publicKey, parentPubKey, and currencySymbol.
 * FR-022: RegistryBinding mandatory for completeness.
 * FR-026: FeePayer is optional.
 * FR-018: Optional ledger attachment.
 *
 * Usage:
 *   const child = new ChildAccountBuilder()
 *     .withPublicKey(childPubKey)
 *     .withParentPublicKey(parentPubKey)
 *     .withCurrency('HBAR')
 *     .withRegistryBinding({ registryIdentifier: '...', externalId: '...' })
 *     .build();
 */
export class ChildAccountBuilder {
  private _publicKey?: NeuronPublicKey | undefined;
  private _parentPubKey?: NeuronPublicKey | undefined;
  private _currencySymbol?: string | undefined;
  private _registryBinding?: RegistryBinding | undefined;
  private _feePayer?: LedgerAccountId | undefined;
  private _ledgerAttachment?: LedgerAttachment | undefined;

  /**
   * Set the Child's secp256k1 public key.
   * FR-007: MUST be a single NeuronPublicKey.
   */
  withPublicKey(pk: NeuronPublicKey): this {
    this._publicKey = pk;
    return this;
  }

  /**
   * Set the reference to the Parent's public key.
   * FR-007: MUST be present for Child accounts.
   */
  withParentPublicKey(pk: NeuronPublicKey): this {
    this._parentPubKey = pk;
    return this;
  }

  /**
   * Set the currency symbol.
   * FR-020: MUST be present on all account types.
   */
  withCurrency(symbol: string): this {
    this._currencySymbol = symbol;
    return this;
  }

  /**
   * Set the registry binding.
   * FR-022: MUST be present for completeness (FR-011a).
   */
  withRegistryBinding(binding: RegistryBinding): this {
    this._registryBinding = binding;
    return this;
  }

  /**
   * Set the fee payer account ID.
   * FR-026: MAY be present on Child accounts.
   */
  withFeePayer(feePayer: LedgerAccountId): this {
    this._feePayer = feePayer;
    return this;
  }

  /**
   * Set the optional ledger attachment.
   * FR-018: MAY be present on any account type.
   */
  withLedgerAttachment(attachment: LedgerAttachment): this {
    this._ledgerAttachment = attachment;
    return this;
  }

  /**
   * Build a Child account with basic validation.
   *
   * FR-011: Validates type-specific rules (V-CHILD-01..02).
   * Throws on the first missing required field.
   *
   * @returns Immutable ChildAccount instance
   * @throws AccountError NEURON-ACCT-002 if publicKey is missing
   * @throws AccountError NEURON-ACCT-002 if parentPubKey is missing
   * @throws AccountError NEURON-ACCT-008 if currencySymbol is missing or empty
   */
  build(): ChildAccount {
    if (this._publicKey == null) {
      throw missingRequiredField('publicKey is required for Child account');
    }
    if (this._parentPubKey == null) {
      throw missingRequiredField('parentPubKey is required for Child account');
    }
    if (this._currencySymbol == null || this._currencySymbol === '') {
      throw invalidCurrencySymbol('Currency symbol is required');
    }

    const account: ChildAccount = {
      accountType: AccountType.Child,
      publicKey: this._publicKey,
      parentPubKey: this._parentPubKey,
      currencySymbol: this._currencySymbol,
      ...(this._registryBinding != null
        ? { registryBinding: this._registryBinding }
        : {}),
      ...(this._feePayer != null
        ? { feePayer: this._feePayer }
        : {}),
      ...(this._ledgerAttachment != null
        ? { ledgerAttachment: this._ledgerAttachment }
        : {}),
    };

    return account;
  }

  /**
   * Build a complete Child account with strict completeness validation.
   *
   * FR-011a: A Child account is complete when it has all MUST fields
   * plus a registry binding.
   *
   * @returns Immutable ChildAccount instance with registry binding
   * @throws AccountError NEURON-ACCT-002 if required fields are missing
   * @throws AccountError NEURON-ACCT-008 if currencySymbol is missing
   * @throws AccountError NEURON-ACCT-007 if registry binding is missing
   */
  buildComplete(): ChildAccount {
    const account = this.build();
    if (account.registryBinding == null) {
      throw accountIncomplete('Child account requires registry binding for completeness');
    }
    return account;
  }
}

/**
 * Fluent builder for Shared accounts.
 *
 * FR-007a: Requires multisigKey and currencySymbol.
 * FR-018: Optional ledger attachment.
 *
 * Usage:
 *   const shared = new SharedAccountBuilder()
 *     .withMultisigKey(multisig)
 *     .withCurrency('HBAR')
 *     .build();
 */
export class SharedAccountBuilder {
  private _multisigKey?: MultisigKey | undefined;
  private _currencySymbol?: string | undefined;
  private _ledgerAttachment?: LedgerAttachment | undefined;

  /**
   * Set the multi-signature key configuration.
   * FR-007a: MUST be present for Shared accounts.
   */
  withMultisigKey(key: MultisigKey): this {
    this._multisigKey = key;
    return this;
  }

  /**
   * Set the currency symbol.
   * FR-020: MUST be present on all account types.
   */
  withCurrency(symbol: string): this {
    this._currencySymbol = symbol;
    return this;
  }

  /**
   * Set the optional ledger attachment.
   * FR-018: MAY be present on any account type.
   */
  withLedgerAttachment(attachment: LedgerAttachment): this {
    this._ledgerAttachment = attachment;
    return this;
  }

  /**
   * Build a Shared account with basic validation.
   *
   * FR-011: Validates type-specific rules (V-SHARED-01).
   * Throws on the first missing required field.
   *
   * @returns Immutable SharedAccount instance
   * @throws AccountError NEURON-ACCT-002 if multisigKey is missing
   * @throws AccountError NEURON-ACCT-008 if currencySymbol is missing or empty
   */
  build(): SharedAccount {
    if (this._multisigKey == null) {
      throw missingRequiredField('multisigKey is required for Shared account');
    }
    if (this._currencySymbol == null || this._currencySymbol === '') {
      throw invalidCurrencySymbol('Currency symbol is required');
    }

    const account: SharedAccount = {
      accountType: AccountType.Shared,
      multisigKey: this._multisigKey,
      currencySymbol: this._currencySymbol,
      ...(this._ledgerAttachment != null
        ? { ledgerAttachment: this._ledgerAttachment }
        : {}),
    };

    return account;
  }

  /**
   * Build a complete Shared account with strict completeness validation.
   *
   * FR-011a: A Shared account is complete when it has all MUST fields
   * plus a ledger attachment.
   *
   * @returns Immutable SharedAccount instance with ledger attachment
   * @throws AccountError NEURON-ACCT-002 if required fields are missing
   * @throws AccountError NEURON-ACCT-008 if currencySymbol is missing
   * @throws AccountError NEURON-ACCT-007 if ledger attachment is missing
   */
  buildComplete(): SharedAccount {
    const account = this.build();
    if (account.ledgerAttachment == null) {
      throw accountIncomplete('Shared account requires ledger attachment for completeness');
    }
    return account;
  }
}

File: impl/golang/internal/account/builder.go
package account

import (
	"github.com/neuron-sdk/neuron-go-sdk/internal/keylib"
)

// isZeroPublicKey reports whether a NeuronPublicKey is the zero value.
func isZeroPublicKey(pk keylib.NeuronPublicKey) bool {
	compressed := pk.Compressed()
	for _, b := range compressed {
		if b != 0 {
			return false
		}
	}
	return true
}

// ---------------------------------------------------------------------------
// ParentAccountBuilder
// ---------------------------------------------------------------------------

// FR-011: Fluent builder API for account construction
// ParentAccountBuilder constructs a Parent NeuronAccount using a fluent API.
// A Parent account requires a public key, a DID, and a currency symbol.
// It MUST NOT have a parentPubKey or multisigKey.
type ParentAccountBuilder struct {
	publicKey        *keylib.NeuronPublicKey
	did              *NeuronDID
	currencySymbol   string
	ledgerAttachment *LedgerAttachment
}

// NewParentAccountBuilder creates a new ParentAccountBuilder.
func NewParentAccountBuilder() *ParentAccountBuilder {
	return &ParentAccountBuilder{}
}

// WithPublicKey sets the public key for the Parent account.
func (b *ParentAccountBuilder) WithPublicKey(pk keylib.NeuronPublicKey) *ParentAccountBuilder {
	b.publicKey = &pk
	return b
}

// WithDID sets the DID for the Parent account.
func (b *ParentAccountBuilder) WithDID(did NeuronDID) *ParentAccountBuilder {
	b.did = &did
	return b
}

// WithCurrency sets the currency symbol for the Parent account.
func (b *ParentAccountBuilder) WithCurrency(symbol string) *ParentAccountBuilder {
	b.currencySymbol = symbol
	return b
}

// WithLedgerAttachment sets the ledger attachment for the Parent account.
func (b *ParentAccountBuilder) WithLedgerAttachment(la LedgerAttachment) *ParentAccountBuilder {
	b.ledgerAttachment = &la
	return b
}

// FR-006: Parent account validation rules
// Build constructs and validates a Parent NeuronAccount.
//
// Validation rules:
//   - V-PARENT-02: publicKey MUST be set
//   - V-PARENT-01: DID MUST be set
//   - V-PARENT-03: currencySymbol MUST be set
//   - V-PARENT-04: parentPubKey MUST NOT be set (enforced by builder: not settable)
//   - V-PARENT-05: multisigKey MUST NOT be set (enforced by builder: not settable)
//
// Derives evmAddress and peerID from publicKey.
func (b *ParentAccountBuilder) Build() (*NeuronAccount, error) {
	// V-PARENT-02: publicKey required
	if b.publicKey == nil || isZeroPublicKey(*b.publicKey) {
		return nil, ValidationError{
			Field:    "publicKey",
			RuleCode: "V-PARENT-02",
			Message:  "Parent account must have a single NeuronPublicKey",
		}
	}

	// V-PARENT-01: DID required
	if b.did == nil || b.did.identifier == "" {
		return nil, ValidationError{
			Field:    "did",
			RuleCode: "V-PARENT-01",
			Message:  "Parent account must have a DID",
		}
	}

	// V-PARENT-03: currency required
	if b.currencySymbol == "" {
		return nil, ValidationError{
			Field:    "currencySymbol",
			RuleCode: "V-PARENT-03",
			Message:  "Parent account must have a currency symbol",
		}
	}

	// Derive EVM address from public key.
	evmAddr := b.publicKey.EVMAddress()

	// Derive PeerID from public key.
	peerID, err := b.publicKey.PeerID()
	if err != nil {
		return nil, ValidationError{
			Field:    "publicKey",
			RuleCode: "V-PARENT-02",
			Message:  "failed to derive PeerID from public key: " + err.Error(),
		}
	}

	acct := &NeuronAccount{
		accountType:    Parent,
		publicKey:      b.publicKey,
		evmAddress:     &evmAddr,
		peerID:         &peerID,
		did:            b.did,
		currencySymbol: b.currencySymbol,
	}

	if b.ledgerAttachment != nil {
		acct.ledgerAttach = b.ledgerAttachment
	}

	return acct, nil
}

// BuildComplete constructs a Parent account and additionally requires that
// a ledger attachment is present.
func (b *ParentAccountBuilder) BuildComplete() (*NeuronAccount, error) {
	acct, err := b.Build()
	if err != nil {
		return nil, err
	}

	if acct.ledgerAttach == nil {
		return nil, ValidationError{
			Field:    "ledgerAttachment",
			RuleCode: "V-PARENT-06",
			Message:  "BuildComplete requires a ledger attachment",
		}
	}

	return acct, nil
}

// ---------------------------------------------------------------------------
// ChildAccountBuilder
// ---------------------------------------------------------------------------

// ChildAccountBuilder constructs a Child NeuronAccount using a fluent API.
// A Child account requires a public key, a parent public key, and a currency symbol.
// It MUST NOT have a DID or multisigKey.
type ChildAccountBuilder struct {
	publicKey        *keylib.NeuronPublicKey
	parentPubKey     *keylib.NeuronPublicKey
	currencySymbol   string
	registryBinding  *RegistryBinding
	feePayer         *LedgerAccountId
	ledgerAttachment *LedgerAttachment
}

// NewChildAccountBuilder creates a new ChildAccountBuilder.
func NewChildAccountBuilder() *ChildAccountBuilder {
	return &ChildAccountBuilder{}
}

// WithPublicKey sets the public key for the Child account.
func (b *ChildAccountBuilder) WithPublicKey(pk keylib.NeuronPublicKey) *ChildAccountBuilder {
	b.publicKey = &pk
	return b
}

// WithParentPublicKey sets the parent's public key for the Child account.
func (b *ChildAccountBuilder) WithParentPublicKey(pk keylib.NeuronPublicKey) *ChildAccountBuilder {
	b.parentPubKey = &pk
	return b
}

// WithCurrency sets the currency symbol for the Child account.
func (b *ChildAccountBuilder) WithCurrency(symbol string) *ChildAccountBuilder {
	b.currencySymbol = symbol
	return b
}

// WithRegistryBinding sets the registry binding for the Child account.
func (b *ChildAccountBuilder) WithRegistryBinding(rb RegistryBinding) *ChildAccountBuilder {
	b.registryBinding = &rb
	return b
}

// WithFeePayer sets the fee payer ledger account ID for the Child account.
func (b *ChildAccountBuilder) WithFeePayer(fp LedgerAccountId) *ChildAccountBuilder {
	b.feePayer = &fp
	return b
}

// WithLedgerAttachment sets the ledger attachment for the Child account.
func (b *ChildAccountBuilder) WithLedgerAttachment(la LedgerAttachment) *ChildAccountBuilder {
	b.ledgerAttachment = &la
	return b
}

// FR-007: Child account validation rules
// Build constructs and validates a Child NeuronAccount.
//
// Validation rules:
//   - V-CHILD-02: publicKey MUST be set
//   - V-CHILD-01: parentPubKey MUST be set
//   - V-CHILD-03: currencySymbol MUST be set
//   - V-CHILD-04: DID MUST NOT be set (enforced by builder: not settable)
//   - V-CHILD-05: multisigKey MUST NOT be set (enforced by builder: not settable)
//
// Derives evmAddress, peerID, and p2pHost (= peerID) from publicKey.
func (b *ChildAccountBuilder) Build() (*NeuronAccount, error) {
	// V-CHILD-02: publicKey required
	if b.publicKey == nil || isZeroPublicKey(*b.publicKey) {
		return nil, ValidationError{
			Field:    "publicKey",
			RuleCode: "V-CHILD-02",
			Message:  "Child account must have a single NeuronPublicKey",
		}
	}

	// V-CHILD-01: parentPubKey required
	if b.parentPubKey == nil || isZeroPublicKey(*b.parentPubKey) {
		return nil, ValidationError{
			Field:    "parentPubKey",
			RuleCode: "V-CHILD-01",
			Message:  "Child account must reference a parent NeuronPublicKey",
		}
	}

	// V-CHILD-03: currency required
	if b.currencySymbol == "" {
		return nil, ValidationError{
			Field:    "currencySymbol",
			RuleCode: "V-CHILD-03",
			Message:  "Child account must have a currency symbol",
		}
	}

	// Derive EVM address from public key.
	evmAddr := b.publicKey.EVMAddress()

	// Derive PeerID from public key.
	peerID, err := b.publicKey.PeerID()
	if err != nil {
		return nil, ValidationError{
			Field:    "publicKey",
			RuleCode: "V-CHILD-02",
			Message:  "failed to derive PeerID from public key: " + err.Error(),
		}
	}

	// p2pHost = peerID for Child accounts.
	p2pHost := peerID

	acct := &NeuronAccount{
		accountType:    Child,
		publicKey:      b.publicKey,
		evmAddress:     &evmAddr,
		peerID:         &peerID,
		parentPubKey:   b.parentPubKey,
		currencySymbol: b.currencySymbol,
		p2pHost:        &p2pHost,
	}

	if b.registryBinding != nil {
		acct.registryBinding = b.registryBinding
	}

	if b.feePayer != nil {
		acct.feePayer = b.feePayer
	}

	if b.ledgerAttachment != nil {
		acct.ledgerAttach = b.ledgerAttachment
	}

	return acct, nil
}

// BuildComplete constructs a Child account and additionally requires that
// both a ledger attachment and a registry binding are present.
func (b *ChildAccountBuilder) BuildComplete() (*NeuronAccount, error) {
	acct, err := b.Build()
	if err != nil {
		return nil, err
	}

	if acct.ledgerAttach == nil {
		return nil, ValidationError{
			Field:    "ledgerAttachment",
			RuleCode: "V-CHILD-06",
			Message:  "BuildComplete requires a ledger attachment",
		}
	}

	if acct.registryBinding == nil {
		return nil, ValidationError{
			Field:    "registryBinding",
			RuleCode: "V-CHILD-07",
			Message:  "BuildComplete requires a registry binding",
		}
	}

	return acct, nil
}

// ---------------------------------------------------------------------------
// SharedAccountBuilder
// ---------------------------------------------------------------------------

// SharedAccountBuilder constructs a Shared NeuronAccount using a fluent API.
// A Shared account requires a multisig key and a currency symbol.
// It MUST NOT have a DID, parentPubKey, or publicKey.
type SharedAccountBuilder struct {
	multisigKey      *keylib.MultisigKey
	currencySymbol   string
	ledgerAttachment *LedgerAttachment
}

// NewSharedAccountBuilder creates a new SharedAccountBuilder.
func NewSharedAccountBuilder() *SharedAccountBuilder {
	return &SharedAccountBuilder{}
}

// WithMultisigKey sets the multisig key for the Shared account.
func (b *SharedAccountBuilder) WithMultisigKey(mk *keylib.MultisigKey) *SharedAccountBuilder {
	b.multisigKey = mk
	return b
}

// WithCurrency sets the currency symbol for the Shared account.
func (b *SharedAccountBuilder) WithCurrency(symbol string) *SharedAccountBuilder {
	b.currencySymbol = symbol
	return b
}

// WithLedgerAttachment sets the ledger attachment for the Shared account.
func (b *SharedAccountBuilder) WithLedgerAttachment(la LedgerAttachment) *SharedAccountBuilder {
	b.ledgerAttachment = &la
	return b
}

// Build constructs and validates a Shared NeuronAccount.
//
// Validation rules:
//   - V-SHARED-01: multisigKey MUST be set
//   - V-SHARED-05: currencySymbol MUST be set
//   - V-SHARED-02: DID MUST NOT be set (enforced by builder: not settable)
//   - V-SHARED-03: parentPubKey MUST NOT be set (enforced by builder: not settable)
//   - V-SHARED-04: publicKey MUST NOT be set (enforced by builder: not settable)
//
// No key derivation is performed for Shared accounts.
func (b *SharedAccountBuilder) Build() (*NeuronAccount, error) {
	// V-SHARED-01: multisigKey required
	if b.multisigKey == nil {
		return nil, ValidationError{
			Field:    "multisigKey",
			RuleCode: "V-SHARED-01",
			Message:  "Shared account must have a MultisigKey with threshold",
		}
	}

	// V-SHARED-05: currency required
	if b.currencySymbol == "" {
		return nil, ValidationError{
			Field:    "currencySymbol",
			RuleCode: "V-SHARED-05",
			Message:  "Shared account must have a currency symbol",
		}
	}

	acct := &NeuronAccount{
		accountType:    Shared,
		multisigKey:    b.multisigKey,
		currencySymbol: b.currencySymbol,
	}

	if b.ledgerAttachment != nil {
		acct.ledgerAttach = b.ledgerAttachment
	}

	return acct, nil
}

// BuildComplete constructs a Shared account and additionally requires that
// a ledger attachment is present.
func (b *SharedAccountBuilder) BuildComplete() (*NeuronAccount, error) {
	acct, err := b.Build()
	if err != nil {
		return nil, err
	}

	if acct.ledgerAttach == nil {
		return nil, ValidationError{
			Field:    "ledgerAttachment",
			RuleCode: "V-SHARED-06",
			Message:  "BuildComplete requires a ledger attachment",
		}
	}

	return acct, nil
}
